frontend/src/pages/CustomerListPage.tsx

```diff
@@ -0,0 +1,130 @@
+import React, { useEffect, useState } from 'react';
+import { ArrowDownCircle, ArrowUpCircle, Search } from 'react-bootstrap-icons';
+import { Trans, useTranslation } from 'react-i18next';
+import { shallowEqual, useDispatch, useSelector } from 'react-redux';
+import { TableCustomerRow } from '../components/TableCustomerRow';
+import { StoreDispatchType } from '../redux/index';
+import { roadmapsActions } from '../redux/roadmaps';
+import { allCustomersSelector } from '../redux/roadmaps/selectors';
+import { Customer } from '../redux/roadmaps/types';
+import { RootState } from '../redux/types';
+import {
+  SortingOrders,
+  SortingTypes,
+  sortCustomers,
+} from '../utils/CustomerUtils';
+import '../shared.scss';
+
+interface TableHeader {
+  label: string;
+  sorting: SortingTypes;
+}
+
+export const CustomerListPage = () => {
+  const { t } = useTranslation();
+  const [searchString, setSearchString] = useState('');
+  const [sortingType, setSortingType] = useState(SortingTypes.NO_SORT);
+  const [sortingOrder, setSortingOrder] = useState(SortingOrders.ASCENDING);
+  const customers = useSelector<RootState, Customer[] | undefined>(
+    allCustomersSelector,
+    shallowEqual,
+  );
+  const dispatch = useDispatch<StoreDispatchType>();
+
+  useEffect(() => {
+    if (!customers) dispatch(roadmapsActions.getCustomers());
+  }, [dispatch, customers]);
+
+  const getRenderCustomerList: () => Customer[] = () => {
+    // Filter, search, sort customers
+    const searched = customers?.filter(({ name }) =>
+      name.toLowerCase().includes(searchString),
+    );
+    return sortCustomers(searched || [], sortingType, sortingOrder);
+  };
+
+  const onSearchChange = (value: string) => {
+    setSearchString(value.toLowerCase());
+  };
+
+  const toggleSortOrder = () => {
+    if (sortingOrder === SortingOrders.ASCENDING) {
+      setSortingOrder(SortingOrders.DESCENDING);
+    } else {
+      setSortingOrder(SortingOrders.ASCENDING);
+    }
+  };
+
+  const onSortingChange = (sorter: SortingTypes) => {
+    if (sorter === sortingType) {
+      toggleSortOrder();
+    } else {
+      setSortingOrder(SortingOrders.ASCENDING);
+    }
+    setSortingType(sorter);
+  };
+
+  const sortingArrow = () =>
+    sortingOrder === SortingOrders.ASCENDING ? (
+      <ArrowUpCircle />
+    ) : (
+      <ArrowDownCircle />
+    );
+
+  const Topbar = () => (
+    <div className="topBar">
+      <div className="searchBarContainer">
+        <input
+          className="search"
+          placeholder={t('Search for customers')}
+          onChange={(e: any) => onSearchChange(e.currentTarget.value)}
+        />
+        <Search />
+      </div>
```

Comment:
Käyttöliittymän läpi testatessa huomasin, että mulla tuo searchtermi katoaa hakupalkista kun kirjoitan siihen jotain.
Esim jos kirjoitan "2" niin haku tuolla termillä toimii, mutta input-kursori ja tuo kirjoittamani 2 katoaa tosta hakupalkista.
Ehkä tossa tapahtuu ton inputin re-render tai jotain joka saa sen syötteen häviämään. Tämä myös estää tuon hakutermin kummaamisen pois. Samalla voisi lisätä ton samanlaisen valkosen palkin listaukseen niin kuin taskilistassa

Reply:
Tutkin itsekin, mistä tuo voisi johtua ja ainoa ero tasklistan toimivaan searchiin on että täällä se renderöidään
`<Topbar/>`
taskilistassa tyyliin
`{Topbar()}`

ja tuon muuttaminen näyttää sen korjaavan, täytynee itsekin tutkia mitä eroa tolla semantiikalla on

Reply:
Ah, eli toi <Topbar/> evaluoituu React.createComponentiksi ja toi alempi vaan suoraan tohon JSX mitä se renderöi, eli toi <Topbar/> syntaksi kadotaa hookkien kontekstin ja muutenkin kun toi customerpage re-renderoityy, se uudelleen definee ton const TopBar funktion ja reactin mielestä <Topbar/> ei ole enää sama <Topbar/> kun aiemmin ja se re-renderaa sen ja input tyhjenee.

toi lähestymistapa toimis esim sillä että samas files ois ton customerpage komponentin ulkopuolella

```
const Topbar = ({ t, onSearchChange }) => (
  <div className="topBar">
    <div className="searchBarContainer">
      <input
        className="search"
        placeholder={t('Search for customers')}
        onChange={(e: any) => onSearchChange(e.currentTarget.value)}
      />
      <Search />
    </div>
  </div>
)


```

Reply:
Laitoin tämän nyt näin: `{Topbar()}`